src/app/clusters/diagnostic-logs-server/diagnostic-logs-server.cpp

```diff
@@ -15,103 +15,274 @@
+        return;
+    }
+
+    Commands::RetrieveLogsResponse::Type response;
+    if (error == CHIP_NO_ERROR)
+    {
+        response.status = StatusEnum::kSuccess;
+        commandHandle->AddResponse(mRequestPath, response);
+    }
+    else
+    {
+        SendErrorResponse(commandHandle, mRequestPath, StatusEnum::kNoLogs);
+    }
+}
+
+void DiagnosticLogsServer::SetAsyncCommandHandleAndPath(CommandHandler * commandObj, const ConcreteCommandPath & commandPath)
+{
+    mAsyncCommandHandle = CommandHandler::Handle(commandObj);
+    mRequestPath        = commandPath;
+}
+
+void DiagnosticLogsServer::SendErrorResponse(chip::app::CommandHandler * commandHandler, chip::app::ConcreteCommandPath path,
+                                             StatusEnum status)
+{
+    Commands::RetrieveLogsResponse::Type response;
+    if (commandHandler != nullptr)
+    {
+        response.status = status;
+        commandHandler->AddResponse(path, response);
+    }
+}
+
+#endif
+
+void DiagnosticLogsServer::HandleLogRequestForResponsePayload(CommandHandler * commandHandler, ConcreteCommandPath path,
+                                                              IntentEnum intent)
+{
+    Commands::RetrieveLogsResponse::Type response;
+    mIntent = intent;
+
+    EndpointId endpoint                       = path.mEndpointId;
+    LogProviderDelegate * logProviderDelegate = GetLogProviderDelegate(endpoint);
+
+    if (isLogProviderDelegateNull(logProviderDelegate, endpoint))
+    {
+        response.status = StatusEnum::kNoLogs;
+        commandHandler->AddResponse(path, response);
+        return;
+    }
+
+    Platform::ScopedMemoryBuffer<uint8_t> buffer;
+
+    if (!buffer.Alloc(kLogContentMaxSize))
+    {
+        ChipLogError(Zcl, "buffer not allocated");
+        response.status = StatusEnum::kNoLogs;
+        commandHandler->AddResponse(path, response);
+        return;
+    }
+
+    mLogSessionHandle = logProviderDelegate->StartLogCollection(intent);
+
+    if (mLogSessionHandle == kInvalidLogSessionHandle)
+    {
+        response.status = StatusEnum::kNoLogs;
+        commandHandler->AddResponse(path, response);
+        return;
+    }
+
+    MutableByteSpan mutableBuffer;
+
+    mutableBuffer = MutableByteSpan(buffer.Get(), kLogContentMaxSize);
+
+    bool isEOF = false;
+
+    // Get the log next chunk
+    uint64_t bytesRead = logProviderDelegate->GetNextChunk(mLogSessionHandle, mutableBuffer, isEOF);
+
+    if (bytesRead == 0)
+    {
+        response.status = StatusEnum::kNoLogs;
+        commandHandler->AddResponse(path, response);
+        return;
+    }
+
+    // log fits. Return in response commandData and call end of log collection
+    if (bytesRead > 0)
+    {
+        if (isEOF)
+        {
+            response.status = StatusEnum::kSuccess;
+        }
+        else
+        {
+            response.status = StatusEnum::kExhausted;
+        }
+        response.logContent = ByteSpan(mutableBuffer.data(), kLogContentMaxSize);
```

Comment:
This code is not exercised yet. its the response payload path. once i add the code to fallback this would be an issue when mutableBuffer.size() < kLogContentMaxSize. Fixed it.